File: python/tvm/contrib/ttile.py
import tvm
from tvm import te


def conv2d(A, W, batch_size, height, width, in_channels, out_channels, kernel_h, kernel_w, pad_h, pad_w, stride_h, stride_w, dilation_h, dilation_w):
    """Create an extern op that compute 2D convolution with Ttile

    Parameters
    ----------
    A: Tensor
        input
    w: Tensor
        convolution weight
    o: Tensor
        output
    height: int
    width: int 
    in_channels: int
    out_channels: int
    kernel_h: int
    kernel_w: int
    pad_h: int
    pad_w: int
    stride_h: int
    stride_w: int
    dilation_h: int
    dilation_w: int
    """
    return te.extern(
    (batch_size, out_channels, (height + 2 * pad_h - kernel_h) // stride_h + 1, (width + 2 * pad_w - kernel_w) // stride_w + 1), #size output
    [A, W],
    lambda ins, outs: tvm.tir.call_packed(
        "tvm.contrib.ttile.conv2d", ins[0], ins[1], outs[0], batch_size, height, width, in_channels, out_channels, kernel_h, kernel_w, pad_h, pad_w, stride_h, stride_w, dilation_h, dilation_w
    ),
    name="Out",
)

File: ttile/conv2d/external_libraries/ttile.c
#undef BIAS

#include <stdio.h>
#include "ttile.h"


void ttile_conv2d(float* K, float * Input, float * Output, int W, int H, int C, int F, int X, int Y) {
	/*
	Simple Convolution

	Inputs:
		- K: weight
		- Input: Input
		- Output: Output
		- W: kernel_w
		- H: kernel_h
		- C: in_channels
		- F: out_channels
		- X: width
		- Y: height

	*/
    for (int f=0; f<F; f++){
        for (int y=0; y<Y; y++){
            for (int x=0; x<X; x++){
                Output[f*Y*X + y*X + x] = 0.0;
                for (int c=0; c<C; c++){
                    for (int h=0; h<H; h++){
                        for (int w=0; w<W; w++){
                            Output[f*Y*X + y*X + x] += Input[c*(Y+H-1)*(X+W-1) + (y+h)*(X+W-1) + x+w] * K[f * C*H*W + c * H*W + h*W + w];
						}
					}
				}
			}
	    }
	}
    return;
}
